metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/sql/DirectSqlTable.java

```diff
@@ -169,56 +176,93 @@ public boolean exists(final QualifiedName name) {
+            final String message = String.format("Originally table %s is not of type iceberg", tableName);
             log.info(message);
-            throw new IllegalStateException(message);
-        } else {
-            updateTableParam(tableId, PARAM_METADATA_LOCK, 1);
+            throw new InvalidMetaException(tableName, message, null);
         }
+        final Map<String, String> newTableMetadata = tableInfo.getMetadata() == null
+            ? Maps.newHashMap() : tableInfo.getMetadata();
+        validateIcebergUpdate(existingTableMetadata, newTableMetadata);
+        final MapDifference<String, String> diff = Maps.difference(existingTableMetadata, newTableMetadata);
+        insertTableParams(tableId, diff.entriesOnlyOnRight());
+        final Map<String, String> updateParams = diff.entriesDiffering().entrySet().stream()
+            .collect(Collectors.toMap(Map.Entry::getKey, s -> s.getValue().rightValue()));
+        updateTableParams(tableId, updateParams);
+        log.debug("Unlocked Iceberg table {}", tableName);
     }
 
-    /**
-     * Unlocks the iceberg table after update so that another request can modify the table.
-     * @param tableId table internal id
-     */
-    public void unlockIcebergTable(final Long tableId) {
-        updateTableParam(tableId, PARAM_METADATA_LOCK, 0);
+    private void validateIcebergUpdate(final Map<String, String> existingTableMetadata,
+                                       final Map<String, String> newTableMetadata) {
+        final String existingMetadataLocation = existingTableMetadata.get(PARAM_METADATA_LOCATION);
+        final String previousMetadataLocation = newTableMetadata.get(PARAM_PREVIOUS_METADATA_LOCATION);
+        final String newMetadataLocation = newTableMetadata.get(DirectSqlTable.PARAM_METADATA_LOCATION);
+        if (StringUtils.isBlank(existingMetadataLocation)) {
+            final String message = "Invalid iceberg table metadata location. Existing metadata location is empty.";
+            log.error(message);
+            throw new IllegalStateException(message);
+        } else if (!Objects.equals(existingMetadataLocation, newMetadataLocation)) {
+            if (StringUtils.isBlank(previousMetadataLocation)) {
+                final String message =
+                    "Invalid iceberg table metadata location. Provided previous metadata location is empty.";
+                log.error(message);
+                throw new IllegalStateException(message);
+            } else if (!Objects.equals(existingMetadataLocation, previousMetadataLocation)) {
+                final String message =
+                    String.format("Invalid iceberg table metadata location (expected:%s, provided:%s)",
+                        existingMetadataLocation, previousMetadataLocation);
+                log.error(message);
+                throw new IllegalStateException(message);
+            }
```

Comment:
If I understand right, the validation logic is 
1. existingTableMetadata is valid ( existingMetadataLocation  not blank..). 
2. newTableMetadata is valid (previousMetadataLocation  not blank )
3. existingMetadataLocation = previousMetadataLocation
The above code will be cleaner if refactor a bit

Reply:
Yes. In this PR, I just moved the code from HIveConnectorFastTableService to this class.

Reply:
Should we make the code better since we are touching them?

Reply:
I feel it is good. If you have any suggestions, let me know.